### PR TITLE
[WIP] add norm for second quantized Hamiltonian

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR adds functions for computing  the 1-norm of a molecular Hamiltonian from the one- and two-electron integrals and eigenvalues of the factorized two-electron integral tensor.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/resources/df.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    .. math::
+
+        T = h_{ij} - \frac{1}{2} \sum_l V_{illj} + \sum_l V_{llij}.
+
+    Note that the two-electron integral tensor must be arranged in the chemist notation.
+
+    Args:
+        one (array[array[float]]): one-electron integrals
+        two (array[array[float]]): two-electron integrals
+        eigvals (array[float]): eigenvalues of the matrices obtained from factorizing the
```
Comment (ixfoduap) [inline pennylane/resources/df.py:56]: Is it really required that we pass the eigenvalues? Can't they just be absorbed inside the integrals?

My understanding of the double factorization is we express V in terms of L's, and the L's are further truncated. But still, we have an equation for V and therefore also the full Hamiltonian. So why can't we compute the norm more directly?

**Review by ixfoduap — CHANGES_REQUESTED** [review]

File: pennylane/resources/__init__.py [hunk 2]
```diff
@@ -14,3 +14,4 @@
 """
 This subpackage provides the functionality for algorithm resource estimation.
 """
+from .df import norm
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/resources/__init__.py]: Changed the module name to `second_quantization` (df was for double factorization).

File: pennylane/resources/df.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for resource estimation with double factorization method.
+"""
+
+from pennylane import numpy as np
+
+
+def norm(one, two, eigvals):
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/resources/df.py]: Changed the function name to `norm_df`. The computation of the norm has method-specific steps for this particular approach. We can still use `norm` though: `qml.SQ.norm`.

File: pennylane/resources/df.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    eigenvalues of the factorized two-electron integral tensor.
+
+    The 1-norm of a double-factorized molecular Hamiltonian is computed as
+    [`arXiv:2007.14460 <https://arxiv.org/abs/2007.14460>`_]
+
+    .. math::
+
+        \lambda = ||T|| + \frac{1}{4} \sum_r ||L^{(r)}||^2,
+
+    where the Schatten norm for a given matrix :math:`T` is defined as
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/resources/df.py]: Modified.

**Review by Jaybsoni — CHANGES_REQUESTED** [review]
Looks good Soran! Just a few quick comments: 

- Is this quantity differentiable? Could we compute the gradient wrt the one and two electron integral parameters? If this doesn't make sense to do then no worries, otherwise it might be worth adding a unit test to check for 
- I imagine this function will not be user facing but mostly used internally in the resource estimation pipeline. If it is user facing, it might be worth adding tests to make sure that if the parameters `one`, `two`, `eigvals` are given in a different framework (autograd or tf arrays) then this function doesn't fail. 
- Codecov is complaining about a few lines which still need to be tested. (you can use `#pragma no cover` to suppress  lines you think have been sufficiently tested but still cause errors in Codcov)

**Review by Jaybsoni — APPROVED** [review]
Looks good !

**Review by ixfoduap — APPROVED** [review]

File: pennylane/resources/second_quantization.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    >>> symbols  = ['H', 'H', 'O']
+    >>> geometry = np.array([[0.00000000,  0.00000000,  0.28377432],
+    >>>                      [0.00000000,  1.45278171, -1.00662237],
+    >>>                      [0.00000000, -1.45278171, -1.00662237]], requires_grad=False)
+    >>> mol = qml.qchem.Molecule(symbols, geometry, basis_name='sto-3g')
+    >>> core, one, two = qml.qchem.electron_integrals(mol)()
+    >>> two = np.swapaxes(two, 1, 3) # convert to the chemists notation
+    >>> _, eigvals, _ = factorize(two, 1e-5)
+    >>> print(norm(one, two, eigvals))
```
Comment (obliviateandsurrender) [inline pennylane/resources/second_quantization.py:72]:
```suggestion
    >>> print(norm_df(one, two, eigvals))
```

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]

File: pennylane/resources/second_quantization.py (hunk 5 above)
Comment (soranjh) [reply pennylane/resources/second_quantization.py:72]: We will use `norm` for all algorithms.

**Review by obliviateandsurrender — APPROVED** [review]
